### ISSUE TITLE
SQL MERGE fails if ON condition is literal FALSE

### ISSUE THREAD

**Issue by djsstarburst** [trinodb/trino#13751]
@przemekak reported that execution of an SQL MERGE statement of this form fails:

```
MERGE INTO target_table t USING source_table s
ON FALSE
WHEN NOT MATCHED THEN INSERT (...) VALUES (...)
```

The failure is due to elimination of the RIGHT JOIN because the join predicate is false.  The fix is to recognize during optimization that the JOIN has been optimized away, and turn the MERGE into an INSERT operation.

**Comment by djsstarburst** [comment]
The fix for the false join condition in MERGE, in PR #13838, is simply removing the call to `rewriteTableScan` when recreating the `MergeWriterNode` with the updated `MergeTarget`.  That call is needed when using `UpdatablePageSource` to route all delete rows to the node that read the file, but SQL MERGE doesn't have that restriction, and doesn't use `UpdatablePageSource`.

**Comment by djsstarburst** [comment]
Removing the `rewriteTableScan` caused the Hive tests to fail because the`HiveTableHandle` in the `MergeTarget` doesn't have a running `AcidTransaction` until `BeginTableWrite.createWriterTarget` calls `ConnectorMetadata.beginMerge`.  The merge handle returned by that method has the `AcidTransaction` initialized, and `rewriteTableScan` is used to update the handle in the `TableScanNode`.

Why Hive care if the table handle in the `TableScanNode` got updated with the `AcidTransaction`?  Two reasons.  First, the code tests in various places that a transaction is running by examining the table handle.  Second, the `AcidTransaction` object contains the `transactionId` with is used in a bunch of places in `SemiTransactionalHiveMetastore`.

Perhaps at some point we can change the Hive transaction object for modify operations to be global and not hidden in the table scan, since AIUI we only allow one modify operation at a time.  But that's a big, potentially dangerous change.

The simplest alternative is to look for the table scan to modify, but not throw an exception if we don't find it for MERGE.  I'm going to try that to verify that it fixes all the tests.

**Comment by djsstarburst** [comment]
This issue was fixed as part of PR #14636, which is now merged.
